internal/storage/metastorage/firestore/block_storage.go

```diff
@@ -0,0 +1,390 @@
+	return b.instrumentPersistBlockMetas.Instrument(ctx, func(ctx context.Context) error {
+		if len(blocks) == 0 {
+			return nil
+		}
+		sort.Slice(blocks, func(i, j int) bool {
+			return blocks[i].Height < blocks[j].Height
+		})
+		if err := parser.ValidateChain(blocks, lastBlock); err != nil {
+			return xerrors.Errorf("failed to validate chain: %w", err)
+		}
+		bulkSize := 0
+		bulkWriter := b.client.BulkWriter(ctx)
+
+		for _, block := range blocks {
+			docRef := b.client.Doc(fmt.Sprintf("env/%s/blocks/%d-%020d-%s", b.env, block.Tag, block.Height, block.Hash))
+			_, err := bulkWriter.Set(docRef, b.fromBlockMetadata(block))
+			if err != nil {
+				return xerrors.Errorf("failed to add block %+v to BulkWriter: %w", block, err)
+			}
+			bulkSize++
+			if bulkSize >= maxBulkWriteSize {
+				bulkWriter.Flush()
+				bulkSize = 0
+			}
+			docRef = b.client.Doc(fmt.Sprintf("env/%s/blocks/canonical-%d-%020d", b.env, block.Tag, block.Height))
+			_, err = bulkWriter.Set(docRef, b.fromBlockMetadata(block))
+			if err != nil {
+				return xerrors.Errorf("failed to add block to BulkWriter: %w", err)
+			}
+			bulkSize++
+			if bulkSize >= maxBulkWriteSize {
+				bulkWriter.Flush()
+				bulkSize = 0
+			}
+		}
+		bulkWriter.End()
+
+		if updateWatermark {
+			watermarkBlock := blocks[len(blocks)-1]
+			docRef := b.client.Doc(fmt.Sprintf("env/%s/blocks/%d-latest", b.env, watermarkBlock.Tag))
+			_, err := docRef.Set(ctx, b.fromBlockMetadata(watermarkBlock))
+			if err != nil {
+				return xerrors.Errorf("failed to update watermark block: %w", err)
+			}
+		}
+
+		return nil
+	})
+}
+
+func (b *blockStorageImpl) validateHeight(height uint64) error {
+	if height < b.blockStartHeight {
+		return xerrors.Errorf(
+			"height(%d) should be no less than blockStartHeight(%d): %w",
+			height, b.blockStartHeight, errors.ErrInvalidHeight)
+	}
+	return nil
+}
+
+func (b *blockStorageImpl) getBlock(ctx context.Context, docName string) (*chainstorage.BlockMetadata, error) {
+	docRef := b.client.Doc(docName)
+	doc, err := docRef.Get(ctx)
+	if err != nil && status.Code(err) != codes.NotFound {
+		return nil, xerrors.Errorf("failed to get block: %w", err)
+	}
+	if !doc.Exists() {
+		return nil, errors.ErrItemNotFound
+	}
+	block, err := b.intoBlockMetadata(doc)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to parse block data: %w", err)
+	}
+	return block, nil
+}
+
+func (*blockStorageImpl) fromBlockMetadata(block *chainstorage.BlockMetadata) map[string]interface{} {
+	v := make(map[string]interface{})
+	v["Hash"] = block.Hash
+	v["ParentHash"] = block.ParentHash
+	v["Height"] = int64(block.Height)
+	v["ParentHeight"] = int64(block.ParentHeight)
+	v["ObjectKeyMain"] = block.ObjectKeyMain
+	v["Skipped"] = block.Skipped
+	v["Tag"] = block.Tag
+	v["Timestamp"] = block.Timestamp
+	return v
+}
+
+func (*blockStorageImpl) intoBlockMetadata(doc *firestore.DocumentSnapshot) (*chainstorage.BlockMetadata, error) {
+	b := &chainstorage.BlockMetadata{}
+	var s struct {
```

Comment:
why do we need to create the s struct here, why not use chainstorage.BlockMetadata?

Reply:
firestore does not support uint64, it has to be converted into int64, which is also enough for storing block height.